File: python/raft/nodo.py
# Nodo Raft en Python, sobre sockets TCP crudos e hilos (socket + threading de la
# stdlib, sin frameworks). Habla EXACTAMENTE el mismo protocolo de texto que el
# nodo Java, asi los nodos de los tres lenguajes forman un solo cluster.
#
# Protocolo (una linea por mensaje, campos separados por |):
#   REQUEST_VOTE|term|candidatoId|ultimoLogIndex|ultimoLogTerm
#   VOTE|term|otorgado(1/0)
#   APPEND|term|liderId|prevLogIndex|prevLogTerm|commitIndex|entradas
#   APPEND_OK|term|exito(1/0)|matchIndex
#   NUEVA_DETECCION|tipo,imagenRef,fechaHora,camara
#   LEER_REGISTRO / REGISTRO|det1;det2;...
#   REDIRECT|host:puerto / OK
#
# "entradas" son  term,comando  separadas por ; (vacio en heartbeats).

import socket
import threading
import time
import random

SEP_ENTRADAS = ";"


class Entrada:
    __slots__ = ("term", "comando")

    def __init__(self, term, comando):
        self.term = term
        self.comando = comando

    def serializar(self):
        return f"{self.term},{self.comando}"

    @staticmethod
    def deserializar(s):
        coma = s.index(",")
        return Entrada(int(s[:coma]), s[coma + 1:])


class NodoRaft:
    def __init__(self, mi_id, puerto, pares):
        # pares: lista de (id, host, puerto) de los OTROS nodos
        self.id = mi_id
        self.puerto = puerto
        self.pares = pares

        # estado persistente
        self.current_term = 0
        self.voted_for = None
        self.log = []  # log[0] es indice 1

        # estado volatil
        self.estado = "SEGUIDOR"
        self.commit_index = 0
        self.last_applied = 0
        self.lider_actual = None

        # estado de lider
        self.next_index = {}
        self.match_index = {}

        # maquina de estado: el registro de detecciones aplicado
        self.registro = []

        # tiempo
        self.ultimo_contacto = 0
        self.timeout_eleccion = self._nuevo_timeout()
        self.HEARTBEAT = 0.05

        self.lock = threading.RLock()
        self.corriendo = True
        self.servidor = None

    def _nuevo_timeout(self):
        return random.uniform(0.150, 0.300)  # 150-300 ms, rango del profe

    def _ahora(self):
        return time.time()

    # ------------------------------------------------------------------
    def iniciar(self):
        self.servidor = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        self.servidor.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        self.servidor.bind(("0.0.0.0", self.puerto))
        self.servidor.listen(50)
        self.ultimo_contacto = self._ahora()
        self._log(f"nodo {self.id} escuchando en puerto {self.puerto} (seguidor)")

        threading.Thread(target=self._bucle_aceptar, daemon=True).start()
        threading.Thread(target=self._bucle_tiempo, daemon=True).start()

    def detener(self):
        self.corriendo = False
        try:
            self.servidor.close()
        except Exception:
            pass

    def _bucle_aceptar(self):
        while self.corriendo:
            try:
                cli, _ = self.servidor.accept()
                threading.Thread(target=self._atender, args=(cli,), daemon=True).start()
            except OSError:
                if self.corriendo:
                    continue
                break

    def _bucle_tiempo(self):
        while self.corriendo:
            time.sleep(0.015)
            if self.estado == "LIDER":
                self._enviar_heartbeats()
                time.sleep(self.HEARTBEAT)
            else:
                if self._ahora() - self.ultimo_contacto >= self.timeout_eleccion:
                    self._iniciar_eleccion()

    # ------------------------------------------------------------------
    # ELECCION
    # ------------------------------------------------------------------
    def _iniciar_eleccion(self):
        with self.lock:
            self.estado = "CANDIDATO"
            self.current_term += 1
            self.voted_for = self.id
            self.lider_actual = None
            self.timeout_eleccion = self._nuevo_timeout()
            self.ultimo_contacto = self._ahora()
            term_eleccion = self.current_term
            ult_idx = len(self.log)
            ult_term = self.log[-1].term if self.log else 0
        self._log(f"inicia eleccion en term {term_eleccion}")

        votos = [1]
        mayoria = (len(self.pares) + 1) // 2 + 1
        hilos = []

        for (pid, host, pto) in self.pares:
            t = threading.Thread(target=self._pedir_voto,
                                 args=(host, pto, term_eleccion, ult_idx, ult_term, votos, mayoria),
                                 daemon=True)
            t.start()
            hilos.append(t)
        for t in hilos:
            t.join(0.2)

    def _pedir_voto(self, host, pto, term_eleccion, ult_idx, ult_term, votos, mayoria):
        resp = self._enviar(host, pto, self._unir("REQUEST_VOTE", term_eleccion, self.id, ult_idx, ult_term))
        if not resp:
            return
        c = resp.split("|")
        if c[0] != "VOTE":
            return
        term_resp = int(c[1])
        otorgado = c[2] == "1"
        with self.lock:
            if term_resp > self.current_term:
                self._volver_seguidor(term_resp)
                return
            if self.estado != "CANDIDATO" or self.current_term != term_eleccion:
                return
            if otorgado:
                votos[0] += 1
                if votos[0] >= mayoria:
                    self._volver_lider()

    def _volver_lider(self):
        if self.estado == "LIDER":
            return
        self.estado = "LIDER"
        self.lider_actual = self.id
        sig = len(self.log) + 1
        for (pid, _, _) in self.pares:
            self.next_index[pid] = sig
            self.match_index[pid] = 0
        self._log(f"*** soy LIDER en term {self.current_term} ***")
        self._enviar_heartbeats()

    def _volver_seguidor(self, term):
        self.current_term = term
        self.estado = "SEGUIDOR"
        self.voted_for = None
        self.ultimo_contacto = self._ahora()

    # ------------------------------------------------------------------
    # REPLICACION
    # ------------------------------------------------------------------
    def _enviar_heartbeats(self):
        for (pid, host, pto) in self.pares:
            threading.Thread(target=self._replicar_a, args=(pid, host, pto), daemon=True).start()

    def _replicar_a(self, pid, host, pto):
        with self.lock:
            if self.estado != "LIDER":
                return
            term = self.current_term
            ni = self.next_index.get(pid, len(self.log) + 1)
            prev_idx = ni - 1
            prev_term = self.log[prev_idx - 1].term if 0 < prev_idx <= len(self.log) else 0
            a_enviar = self.log[ni - 1:]
            commit = self.commit_index

        ent = SEP_ENTRADAS.join(e.serializar() for e in a_enviar)
        resp = self._enviar(host, pto, self._unir("APPEND", term, self.id, prev_idx, prev_term, commit, ent))
        if not resp:
            return
        c = resp.split("|")
        if c[0] != "APPEND_OK":
            return
        term_resp = int(c[1])
        exito = c[2] == "1"
        match = int(c[3])

        with self.lock:
            if term_resp > self.current_term:
                self._volver_seguidor(term_resp)
                return
            if self.estado != "LIDER" or term != self.current_term:
                return
            if exito:
                self.match_index[pid] = match
                self.next_index[pid] = match + 1
                self._recalcular_commit()
            else:
                ni = self.next_index.get(pid, len(self.log) + 1)
                self.next_index[pid] = max(1, ni - 1)

    def _recalcular_commit(self):
        for n in range(len(self.log), self.commit_index, -1):
            if self.log[n - 1].term != self.current_term:
                continue
            cuenta = 1
            for (pid, _, _) in self.pares:
                if self.match_index.get(pid, 0) >= n:
                    cuenta += 1
            if cuenta >= (len(self.pares) + 1) // 2 + 1:
                self.commit_index = n
                self._aplicar()
                break

    def _aplicar(self):
        while self.last_applied < self.commit_index:
            self.last_applied += 1
            cmd = self.log[self.last_applied - 1].comando
            self.registro.append(cmd)
            self._log(f"aplica[{self.last_applied}]: {cmd}")

    # ------------------------------------------------------------------
    # ATENCION DE MENSAJES
    # ------------------------------------------------------------------
    def _atender(self, cli):
        try:
            f = cli.makefile("rwb")
            linea = f.readline().decode("utf-8").rstrip("\n")
            if not linea:
                return
            resp = self._procesar(linea)
            if resp is not None:
                f.write((resp + "\n").encode("utf-8"))
                f.flush()
        except Exception:
            pass
        finally:
            try:
                cli.close()
            except Exception:
                pass

    def _procesar(self, linea):
        c = linea.split("|")
        tipo = c[0]
        if tipo == "REQUEST_VOTE":
            return self._on_request_vote(c)
        if tipo == "APPEND":
            return self._on_append(c)
        if tipo == "NUEVA_DETECCION":
            return self._on_nueva_deteccion(c)
        if tipo == "LEER_REGISTRO":
            return self._on_leer_registro()
        return None

    def _on_request_vote(self, c):
        term = int(c[1]); cand = int(c[2]); cand_idx = int(c[3]); cand_term = int(c[4])
        with self.lock:
            if term > self.current_term:
                self._volver_seguidor(term)
            otorgar = False
            if term >= self.current_term and (self.voted_for is None or self.voted_for == cand):
                mi_idx = len(self.log)
                mi_term = self.log[-1].term if self.log else 0
                al_menos = cand_term > mi_term or (cand_term == mi_term and cand_idx >= mi_idx)
                if al_menos:
                    otorgar = True
                    self.voted_for = cand
                    self.ultimo_contacto = self._ahora()
            return self._unir("VOTE", self.current_term, 1 if otorgar else 0)

    def _on_append(self, c):
        term = int(c[1]); lider = int(c[2]); prev_idx = int(c[3])
        prev_term = int(c[4]); commit_lider = int(c[5]); entradas_str = c[6]
        with self.lock:
            if term < self.current_term:
                return self._unir("APPEND_OK", self.current_term, 0, 0)
            if term > self.current_term:
                self._volver_seguidor(term)
            self.estado = "SEGUIDOR"
            self.lider_actual = lider
            self.ultimo_contacto = self._ahora()

            if prev_idx > 0:
                if len(self.log) < prev_idx or self.log[prev_idx - 1].term != prev_term:
                    return self._unir("APPEND_OK", self.current_term, 0, 0)

            if entradas_str:
                items = entradas_str.split(SEP_ENTRADAS)
                idx = prev_idx
                for it in items:
                    idx += 1
                    e = Entrada.deserializar(it)
                    if len(self.log) >= idx:
                        if self.log[idx - 1].term != e.term:
                            del self.log[idx - 1:]
                            self.log.append(e)
                    else:
                        self.log.append(e)

            if commit_lider > self.commit_index:
                self.commit_index = min(commit_lider, len(self.log))
                self._aplicar()
            return self._unir("APPEND_OK", self.current_term, 1, len(self.log))

    def _on_nueva_deteccion(self, c):
        with self.lock:
            if self.estado != "LIDER":
                if self.lider_actual is not None:
                    d = self._dir_de(self.lider_actual)
                    if d:
                        return self._unir("REDIRECT", d)
                return self._unir("REDIRECT", "?")
            self.log.append(Entrada(self.current_term, c[1]))
            self._log(f"cliente inserta: {c[1]} (indice {len(self.log)})")
        self._enviar_heartbeats()
        return "OK"

    def _on_leer_registro(self):
        with self.lock:
            if self.estado != "LIDER":
                if self.lider_actual is not None:
                    d = self._dir_de(self.lider_actual)
                    if d:
                        return self._unir("REDIRECT", d)
                return self._unir("REDIRECT", "?")
            return self._unir("REGISTRO", SEP_ENTRADAS.join(self.registro))

    def _dir_de(self, idb):
        if idb == self.id:
            return f"127.0.0.1:{self.puerto}"
        for (pid, host, pto) in self.pares:
            if pid == idb:
                return f"{host}:{pto}"
        return None

    # ------------------------------------------------------------------
    def _enviar(self, host, pto, mensaje):
        try:
            s = socket.create_connection((host, pto), timeout=0.15)
            s.settimeout(0.25)
            f = s.makefile("rwb")
            f.write((mensaje + "\n").encode("utf-8"))
            f.flush()
            resp = f.readline().decode("utf-8").rstrip("\n")
            s.close()
            return resp if resp else None
        except Exception:
            return None

    def _unir(self, *campos):
        return "|".join(str(x) for x in campos)

    def _log(self, m):
        print(f"[py-nodo {self.id}] {m}", flush=True)

File: go/raft/nodo.go
// Package raft: nodo Raft en Go sobre sockets TCP crudos y goroutines (net +
// sync de la stdlib, sin frameworks). Habla el MISMO protocolo de texto que los
// nodos Java y Python, asi los tres forman un solo cluster heterogeneo.
//
// Protocolo (una linea por mensaje, campos separados por |):
//
//	REQUEST_VOTE|term|candidatoId|ultimoLogIndex|ultimoLogTerm
//	VOTE|term|otorgado(1/0)
//	APPEND|term|liderId|prevLogIndex|prevLogTerm|commitIndex|entradas
//	APPEND_OK|term|exito(1/0)|matchIndex
//	NUEVA_DETECCION|tipo,imagenRef,fechaHora,camara
//	LEER_REGISTRO / REGISTRO|det1;det2;...
//	REDIRECT|host:puerto / OK
package raft

import (
	"bufio"
	"fmt"
	"math/rand"
	"net"
	"strconv"
	"strings"
	"sync"
	"time"
)

const sepEntradas = ";"

// Entrada del log replicado.
type Entrada struct {
	Term    int
	Comando string
}

func (e Entrada) serializar() string {
	return strconv.Itoa(e.Term) + "," + e.Comando
}

func deserializar(s string) Entrada {
	coma := strings.Index(s, ",")
	t, _ := strconv.Atoi(s[:coma])
	return Entrada{Term: t, Comando: s[coma+1:]}
}

// Par: un nodo del cluster.
type Par struct {
	ID     int
	Host   string
	Puerto int
}

func (p Par) dir() string { return fmt.Sprintf("%s:%d", p.Host, p.Puerto) }

// Nodo Raft.
type Nodo struct {
	id     int
	puerto int
	pares  []Par

	mu          sync.Mutex
	currentTerm int
	votedFor    int // -1 = nadie
	log         []Entrada

	estado         string // SEGUIDOR, CANDIDATO, LIDER
	commitIndex    int
	lastApplied    int
	liderActual    int // -1 = ninguno
	nextIndex      map[int]int
	matchIndex     map[int]int
	registro       []string
	ultimoContacto time.Time
	timeoutElec    time.Duration

	corriendo bool
	ln        net.Listener
}

func NuevoNodo(id, puerto int, pares []Par) *Nodo {
	return &Nodo{
		id: id, puerto: puerto, pares: pares,
		votedFor: -1, liderActual: -1, estado: "SEGUIDOR",
		nextIndex: map[int]int{}, matchIndex: map[int]int{},
		corriendo: true, timeoutElec: nuevoTimeout(),
	}
}

func nuevoTimeout() time.Duration {
	return time.Duration(150+rand.Intn(151)) * time.Millisecond // 150-300 ms
}

// Iniciar arranca el listener y el bucle de tiempo.
func (n *Nodo) Iniciar() error {
	ln, err := net.Listen("tcp", fmt.Sprintf("0.0.0.0:%d", n.puerto))
	if err != nil {
		return err
	}
	n.ln = ln
	n.ultimoContacto = time.Now()
	n.logf("nodo %d escuchando en puerto %d (seguidor)", n.id, n.puerto)
	go n.bucleAceptar()
	go n.bucleTiempo()
	return nil
}

func (n *Nodo) Detener() {
	n.mu.Lock()
	n.corriendo = false
	n.mu.Unlock()
	if n.ln != nil {
		n.ln.Close()
	}
}

func (n *Nodo) bucleAceptar() {
	for {
		c, err := n.ln.Accept()
		if err != nil {
			n.mu.Lock()
			corriendo := n.corriendo
			n.mu.Unlock()
			if !corriendo {
				return
			}
			continue
		}
		go n.atender(c)
	}
}

func (n *Nodo) bucleTiempo() {
	for {
		time.Sleep(15 * time.Millisecond)
		n.mu.Lock()
		if !n.corriendo {
			n.mu.Unlock()
			return
		}
		est := n.estado
		venc := time.Since(n.ultimoContacto) >= n.timeoutElec
		n.mu.Unlock()

		if est == "LIDER" {
			n.enviarHeartbeats()
			time.Sleep(50 * time.Millisecond)
		} else if venc {
			n.iniciarEleccion()
		}
	}
}

// ------------------------------------------------------------------ ELECCION
func (n *Nodo) iniciarEleccion() {
	n.mu.Lock()
	n.estado = "CANDIDATO"
	n.currentTerm++
	n.votedFor = n.id
	n.liderActual = -1
	n.timeoutElec = nuevoTimeout()
	n.ultimoContacto = time.Now()
	termElec := n.currentTerm
	ultIdx := len(n.log)
	ultTerm := 0
	if ultIdx > 0 {
		ultTerm = n.log[ultIdx-1].Term
	}
	n.mu.Unlock()
	n.logf("inicia eleccion en term %d", termElec)

	var votos int32 = 1
	mayoria := (len(n.pares)+1)/2 + 1
	var wg sync.WaitGroup
	var vmu sync.Mutex

	for _, p := range n.pares {
		wg.Add(1)
		go func(p Par) {
			defer wg.Done()
			resp := n.enviar(p, unir("REQUEST_VOTE", termElec, n.id, ultIdx, ultTerm))
			if resp == "" {
				return
			}
			c := strings.Split(resp, "|")
			if c[0] != "VOTE" {
				return
			}
			termResp, _ := strconv.Atoi(c[1])
			otorgado := c[2] == "1"
			n.mu.Lock()
			defer n.mu.Unlock()
			if termResp > n.currentTerm {
				n.volverSeguidor(termResp)
				return
			}
			if n.estado != "CANDIDATO" || n.currentTerm != termElec {
				return
			}
			if otorgado {
				vmu.Lock()
				votos++
				alcanzo := int(votos) >= mayoria
				vmu.Unlock()
				if alcanzo {
					n.volverLider()
				}
			}
		}(p)
	}
	esperarConTimeout(&wg, 200*time.Millisecond)
}

func (n *Nodo) volverLider() {
	if n.estado == "LIDER" {
		return
	}
	n.estado = "LIDER"
	n.liderActual = n.id
	sig := len(n.log) + 1
	for _, p := range n.pares {
		n.nextIndex[p.ID] = sig
		n.matchIndex[p.ID] = 0
	}
	n.logf("*** soy LIDER en term %d ***", n.currentTerm)
	go n.enviarHeartbeats()
}

func (n *Nodo) volverSeguidor(term int) {
	n.currentTerm = term
	n.estado = "SEGUIDOR"
	n.votedFor = -1
	n.ultimoContacto = time.Now()
}

// -------------------------------------------------------------- REPLICACION
func (n *Nodo) enviarHeartbeats() {
	for _, p := range n.pares {
		go n.replicarA(p)
	}
}

func (n *Nodo) replicarA(p Par) {
	n.mu.Lock()
	if n.estado != "LIDER" {
		n.mu.Unlock()
		return
	}
	term := n.currentTerm
	ni := n.nextIndex[p.ID]
	if ni == 0 {
		ni = len(n.log) + 1
	}
	prevIdx := ni - 1
	prevTerm := 0
	if prevIdx > 0 && prevIdx <= len(n.log) {
		prevTerm = n.log[prevIdx-1].Term
	}
	var aEnviar []Entrada
	if ni <= len(n.log) {
		aEnviar = append(aEnviar, n.log[ni-1:]...)
	}
	commit := n.commitIndex
	n.mu.Unlock()

	partes := make([]string, len(aEnviar))
	for i, e := range aEnviar {
		partes[i] = e.serializar()
	}
	ent := strings.Join(partes, sepEntradas)

	resp := n.enviar(p, unir("APPEND", term, n.id, prevIdx, prevTerm, commit, ent))
	if resp == "" {
		return
	}
	c := strings.Split(resp, "|")
	if c[0] != "APPEND_OK" {
		return
	}
	termResp, _ := strconv.Atoi(c[1])
	exito := c[2] == "1"
	match, _ := strconv.Atoi(c[3])

	n.mu.Lock()
	defer n.mu.Unlock()
	if termResp > n.currentTerm {
		n.volverSeguidor(termResp)
		return
	}
	if n.estado != "LIDER" || term != n.currentTerm {
		return
	}
	if exito {
		n.matchIndex[p.ID] = match
		n.nextIndex[p.ID] = match + 1
		n.recalcularCommit()
	} else {
		if n.nextIndex[p.ID] > 1 {
			n.nextIndex[p.ID]--
		}
	}
}

func (n *Nodo) recalcularCommit() {
	for idx := len(n.log); idx > n.commitIndex; idx-- {
		if n.log[idx-1].Term != n.currentTerm {
			continue
		}
		cuenta := 1
		for _, p := range n.pares {
			if n.matchIndex[p.ID] >= idx {
				cuenta++
			}
		}
		if cuenta >= (len(n.pares)+1)/2+1 {
			n.commitIndex = idx
			n.aplicar()
			break
		}
	}
}

func (n *Nodo) aplicar() {
	for n.lastApplied < n.commitIndex {
		n.lastApplied++
		cmd := n.log[n.lastApplied-1].Comando
		n.registro = append(n.registro, cmd)
		n.logf("aplica[%d]: %s", n.lastApplied, cmd)
	}
}

// ------------------------------------------------------------- MENSAJES
func (n *Nodo) atender(c net.Conn) {
	defer c.Close()
	r := bufio.NewReader(c)
	linea, err := r.ReadString('\n')
	if err != nil && linea == "" {
		return
	}
	linea = strings.TrimRight(linea, "\n")
	if linea == "" {
		return
	}
	resp := n.procesar(linea)
	if resp != "" {
		c.Write([]byte(resp + "\n"))
	}
}

func (n *Nodo) procesar(linea string) string {
	c := strings.Split(linea, "|")
	switch c[0] {
	case "REQUEST_VOTE":
		return n.onRequestVote(c)
	case "APPEND":
		return n.onAppend(c)
	case "NUEVA_DETECCION":
		return n.onNuevaDeteccion(c)
	case "LEER_REGISTRO":
		return n.onLeerRegistro()
	}
	return ""
}

func (n *Nodo) onRequestVote(c []string) string {
	term, _ := strconv.Atoi(c[1])
	cand, _ := strconv.Atoi(c[2])
	candIdx, _ := strconv.Atoi(c[3])
	candTerm, _ := strconv.Atoi(c[4])
	n.mu.Lock()
	defer n.mu.Unlock()
	if term > n.currentTerm {
		n.volverSeguidor(term)
	}
	otorgar := false
	if term >= n.currentTerm && (n.votedFor == -1 || n.votedFor == cand) {
		miIdx := len(n.log)
		miTerm := 0
		if miIdx > 0 {
			miTerm = n.log[miIdx-1].Term
		}
		alMenos := candTerm > miTerm || (candTerm == miTerm && candIdx >= miIdx)
		if alMenos {
			otorgar = true
			n.votedFor = cand
			n.ultimoContacto = time.Now()
		}
	}
	g := 0
	if otorgar {
		g = 1
	}
	return unir("VOTE", n.currentTerm, g)
}

func (n *Nodo) onAppend(c []string) string {
	term, _ := strconv.Atoi(c[1])
	lider, _ := strconv.Atoi(c[2])
	prevIdx, _ := strconv.Atoi(c[3])
	prevTerm, _ := strconv.Atoi(c[4])
	commitLider, _ := strconv.Atoi(c[5])
	entradasStr := c[6]

	n.mu.Lock()
	defer n.mu.Unlock()
	if term < n.currentTerm {
		return unir("APPEND_OK", n.currentTerm, 0, 0)
	}
	if term > n.currentTerm {
		n.volverSeguidor(term)
	}
	n.estado = "SEGUIDOR"
	n.liderActual = lider
	n.ultimoContacto = time.Now()

	if prevIdx > 0 {
		if len(n.log) < prevIdx || n.log[prevIdx-1].Term != prevTerm {
			return unir("APPEND_OK", n.currentTerm, 0, 0)
		}
	}

	if entradasStr != "" {
		items := strings.Split(entradasStr, sepEntradas)
		idx := prevIdx
		for _, it := range items {
			idx++
			e := deserializar(it)
			if len(n.log) >= idx {
				if n.log[idx-1].Term != e.Term {
					n.log = n.log[:idx-1]
					n.log = append(n.log, e)
				}
			} else {
				n.log = append(n.log, e)
			}
		}
	}

	if commitLider > n.commitIndex {
		if commitLider < len(n.log) {
			n.commitIndex = commitLider
		} else {
			n.commitIndex = len(n.log)
		}
		n.aplicar()
	}
	return unir("APPEND_OK", n.currentTerm, 1, len(n.log))
}

func (n *Nodo) onNuevaDeteccion(c []string) string {
	n.mu.Lock()
	if n.estado != "LIDER" {
		d := n.dirDe(n.liderActual)
		n.mu.Unlock()
		if d != "" {
			return unir("REDIRECT", d)
		}
		return unir("REDIRECT", "?")
	}
	n.log = append(n.log, Entrada{Term: n.currentTerm, Comando: c[1]})
	n.logf("cliente inserta: %s (indice %d)", c[1], len(n.log))
	n.mu.Unlock()
	n.enviarHeartbeats()
	return "OK"
}

func (n *Nodo) onLeerRegistro() string {
	n.mu.Lock()
	defer n.mu.Unlock()
	if n.estado != "LIDER" {
		d := n.dirDe(n.liderActual)
		if d != "" {
			return unir("REDIRECT", d)
		}
		return unir("REDIRECT", "?")
	}
	return unir("REGISTRO", strings.Join(n.registro, sepEntradas))
}

func (n *Nodo) dirDe(idb int) string {
	if idb == -1 {
		return ""
	}
	if idb == n.id {
		return fmt.Sprintf("127.0.0.1:%d", n.puerto)
	}
	for _, p := range n.pares {
		if p.ID == idb {
			return p.dir()
		}
	}
	return ""
}

// Estado / Registro exportan lo minimo para pruebas.
func (n *Nodo) Estado() string {
	n.mu.Lock()
	defer n.mu.Unlock()
	return n.estado
}

func (n *Nodo) Term() int {
	n.mu.Lock()
	defer n.mu.Unlock()
	return n.currentTerm
}

func (n *Nodo) Registro() []string {
	n.mu.Lock()
	defer n.mu.Unlock()
	return append([]string{}, n.registro...)
}

func (n *Nodo) Corriendo() bool {
	n.mu.Lock()
	defer n.mu.Unlock()
	return n.corriendo
}

// ------------------------------------------------------------- red util
func (n *Nodo) enviar(p Par, mensaje string) string {
	c, err := net.DialTimeout("tcp", p.dir(), 120*time.Millisecond)
	if err != nil {
		return ""
	}
	defer c.Close()
	c.SetDeadline(time.Now().Add(250 * time.Millisecond))
	c.Write([]byte(mensaje + "\n"))
	r := bufio.NewReader(c)
	resp, _ := r.ReadString('\n')
	return strings.TrimRight(resp, "\n")
}

func unir(campos ...interface{}) string {
	partes := make([]string, len(campos))
	for i, c := range campos {
		partes[i] = fmt.Sprintf("%v", c)
	}
	return strings.Join(partes, "|")
}

func esperarConTimeout(wg *sync.WaitGroup, d time.Duration) {
	done := make(chan struct{})
	go func() { wg.Wait(); close(done) }()
	select {
	case <-done:
	case <-time.After(d):
	}
}

func (n *Nodo) logf(f string, a ...interface{}) {
	fmt.Printf("[go-nodo %d] ", n.id)
	fmt.Printf(f+"\n", a...)
}
